### PR TITLE
Send exact target where the fingerprint was found.

### PR DESCRIPTION
**Pull request by deadly-panda**
With a sprinkle of refactoring!

### CODE REVIEW

**Review by 3asm — CHANGES_REQUESTED** [review]

File: agent/whatweb_agent.py [hunk 1]
```diff
@@ -264,27 +275,47 @@ def _parse_emit_result(self, name: str, output_file: io.BytesIO,
...
+                                    target: DomainTarget | IPTarget) -> agent_report_vulnerability_mixin.AssetLocation:
+        """Returns the target data where the fingerprint was found."""
+        metadata_type = agent_report_vulnerability_mixin.MetaDataType.PORT
+        metadata_value = str(target.port)
+        metadata = [
+            agent_report_vulnerability_mixin.AssetLocationMetaData(type=metadata_type, value=metadata_value)
+        ]
+        if isinstance(target, DomainTarget):
+            asset = domain_asset.DomainName(name=target.name)
+            return agent_report_vulnerability_mixin.AssetLocation(asset=asset, metadata=metadata)
```
[the remark this answers is not in the document]
Comment (deadly-panda) [reply agent/whatweb_agent.py:288]: That's exactly why this PR is a draft 😄

File: agent/whatweb_agent.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
 import subprocess
 import ipaddress
 import tempfile
-from typing import List, Optional, Dict, Any
+from typing import List, Optional, Dict, Any, Union
```
Comment (3asm) [inline agent/whatweb_agent.py:8]: Use | instead of union.

File: agent/whatweb_agent.py [hunk 3]
```diff
@@ -89,14 +100,14 @@ def process(self, message: msg.Message) -> None:
...
                 with tempfile.NamedTemporaryFile() as fp:
-                    self._start_scan(target.name, fp.name)
-                    self._parse_emit_result(target.name, io.BytesIO(fp.read()), target.port, target.schema)
+                    self._start_scan(target, fp.name)
+                    self._parse_emit_result(target, io.BytesIO(fp.read()))
             except subprocess.CalledProcessError as e:
                 logger.error(e)
 
-    def _prepare_targets(self, message: msg.Message) -> List[Target]:
+    def _prepare_targets(self, message: msg.Message) -> List[Union[DomainTarget, IPTarget]]:
```
Comment (3asm) [inline agent/whatweb_agent.py:108]: | instead of union as we support 3.10.

File: agent/whatweb_agent.py [hunk 4]
```diff
@@ -323,32 +355,33 @@ def _send_detected_fingerprints(self, name: str, port: Optional[int] = None, sch
...
                 ),
                 technical_detail=f'Found library `{library_name}` of type '
-                                 f'`{fingerprint_type}` in target `{name}`',
-                risk_rating=agent_report_vulnerability_mixin.RiskRating.INFO)
+                                 f'`{fingerprint_type}` in target `{target.name}`',
+                risk_rating=vuln_mixin.RiskRating.INFO,
+                vulnerability_location=vulnerable_target_data)
 
-    def _get_msg_data(self, name: str, port: Optional[int] = None, schema: Optional[str] = None,
+    def _get_msg_data(self, target: DomainTarget | IPTarget,
```
Comment (3asm) [inline agent/whatweb_agent.py:362]: See, that's inconsistent in a single file!

File: tests/whatweb_test.py [hunk 5]
```diff
@@ -224,3 +224,31 @@ def testWhatWebAgent_whenWhatWebReturnsError_ContinueProcessing(agent_mock: List
...
+                'ipv4': {
+                    'host': '192.168.0.76',
+                    'mask': '32',
+                    'version': 4
+                    },
+                'metadata': [{
+                    'value': '443',
+                    'type': 'PORT'
+                }]}
+            assert any(fingerprint.data.get('vulnerability_location') == fp_location for fingerprint in agent_mock)
```
Comment (3asm) [inline tests/whatweb_test.py:253]: Missing unit test for domain name.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: agent/whatweb_agent.py [hunk 6]
```diff
@@ -264,27 +275,48 @@ def _parse_emit_result(self, name: str, output_file: io.BytesIO,
...
+                                    target: DomainTarget | IPTarget) -> vuln_mixin.VulnerabilityLocation:
+        """Returns the target data where the fingerprint was found."""
+        metadata_type = vuln_mixin.MetadataType.PORT
+        metadata_value = str(target.port)
+        metadata = [
+            vuln_mixin.VulnerabilityLocationMetadata(type=metadata_type, value=metadata_value)
+        ]
+        if isinstance(target, DomainTarget):
+            asset = domain_asset.DomainName(name=target.name)
+            return vuln_mixin.VulnerabilityLocation(asset=asset, metadata=metadata)
```
Comment (3asm) [inline agent/whatweb_agent.py:288]: Shouldn't this be URL as metadata for domain.

**Review by 3asm — CHANGES_REQUESTED** [review]

**Pull request merged**